### ISSUE TITLE
Bug with trailing return types for c++ source code

### ISSUE THREAD

**Issue by nicokoe** [michaeljones/breathe#441]
It seems like trailing return types in c++ cannot get parsed correctly by breathe. Here is an example:

Original source code including a trailing return type:

```
class Thingy;

//! \brief Function that creates a thingy.
auto MakeThingy() -> Thingy*;
```

The generated xml output of doxygen:
```
      <sectiondef kind="func">
      <memberdef kind="function" id="id" prot="public" static="no" const="no" explicit="no" inline="no" virt="non-virtual">
        <type>auto</type>
        <definition>auto MakeThingy</definition>
        <argsstring>() -&gt; Thingy *</argsstring>
        <name>MakeThingy</name>
        <briefdescription>
<para>Function that creates a thingy. </para>
        </briefdescription>
        <detaileddescription>
        </detaileddescription>
        <inbodydescription>
        </inbodydescription>
        <location file=..." line="17" column="1"/>
      </memberdef>
      </sectiondef>
```
The index.rst for sphinx:

```
	.. doxygenfunction:: MakeThingy
```

And the final index.html generated with sphinx and breathe:

![makethingy](https://user-images.githubusercontent.com/27962036/59491610-6489ea80-8e87-11e9-8ba4-51c037db7986.PNG)

Unfortunately, the final index.html is missing the trailing return type which is essential for understanding the function's signature. Doxygen itself seems to be able to handle the trailing return types somehow.

Is there a possibility for you to fix this issue?

**Comment by Sonicsystem** [comment]
I would really like to use the doxygen, sphinx and breathe combination in my project. But this problem comes with major drawbacks and it is not feasible for me to change the source code.

**Comment by t-b** [comment]
@Sonicsystem As you can see on the issues vs PR numbers this project lacks developers. So either start diving into the code and fix it or wait. Unfortuately this is how it is.

**Comment by vermeeren** [comment]
With #512 merged in a bit this now only (may) need support upstream in Sphinx, so said PR will close this issue.

**Comment by vermeeren** [comment]
Should be fixed with #512, released in Breathe v4.17.0. Note that you also need Sphinx 3.x for recent Breathe versions.

**Comment by jakobandersen** [comment]
And it is now fixed on the Sphinx side in the 3.x branch (future v3.1).

**Comment by nicokoe** [comment]
@jakobandersen & @vermeeren: It's been a long time for me and I'm not really into the subject anymore, but a little test I just did on myself on this topic still doesn't show the expected result for "trailing return types". See log and result:
![docs-warning](https://user-images.githubusercontent.com/27962036/83446856-bc7e8000-a44f-11ea-9315-5d580cc03fe2.PNG)
![trailing-return-type-bug](https://user-images.githubusercontent.com/27962036/83446862-c0120700-a44f-11ea-9080-a55407ddf9f5.PNG)

The log seems to say that the C++ code was not read correctly and the result does not show the expected "auto":
auto thingy::MakeThingy () -> Thingy *

But we can still keep this closed until someone else wants it to be fixed :)

**Comment by jakobandersen** [comment]
Which Breathe version are you using?

**Comment by nicokoe** [comment]
Breathe v4.18.1 and Sphinx v3.0.4. Maybe this just works with a future release of sphinx?

**Comment by jakobandersen** [comment]
Yes, you need Sphinx branch 3.x, which will be released as v3.1 (soon if I remember correctly).
